File: perl/math/sieve.pl
#!/usr/bin/env perl
# Print the primes below 100, sieved over an array of flags indexed by the number itself.

use strict;
use warnings;

my $limit = 100;
my @is_prime = (1) x $limit;
@is_prime[0, 1] = (0, 0);

for my $n (2 .. int sqrt $limit) {
    next unless $is_prime[$n];
    for (my $multiple = $n * $n; $multiple < $limit; $multiple += $n) {
        $is_prime[$multiple] = 0;
    }
}

print join(' ', grep { $is_prime[$_] } 0 .. $limit - 1), "\n";

File: ruby/math/sieve.rb
#!/usr/bin/env ruby
# Print the primes below 100, sieved over an array of flags stepped by each prime found.

LIMIT = 100

is_prime = Array.new(LIMIT, true)
is_prime[0] = is_prime[1] = false

(2..Integer.sqrt(LIMIT)).each do |n|
  next unless is_prime[n]

  (n * n).step(LIMIT - 1, n) { |multiple| is_prime[multiple] = false }
end

puts is_prime.each_index.select { |n| is_prime[n] }.join(' ')
